Add command for k8sd recovery after quorum loss (#618)
canonical/k8s-snap: docs/src/snap/howto/restore-quorum.md

replicated on two or more secondary nodes.

However, if a majority of the nodes are lost, the cluster will become unavaialble.
If at least one database node survived, the cluster can be recovered using the
steps outlined in this document.

*Note*: this guide can be used to recover the default Canonical Kubernetes datastore,
dqlite. Persistent Volumes residing on the lost nodes are *not* recovered.

## Dqlite configuration

Be aware that Canonical Kubernetes uses not one, but two dqlite databases:

* k8s-dqlite - used by Kubernetes itself
* k8sd - Kuberenetes cluster management data

Each database has its own state directory:

* ``/var/snap/k8s/common/var/lib/k8s-dqlite``
* ``/var/snap/k8s/common/var/lib/k8sd/state``

The state directory normally contains:

* ``info.yaml`` - the id, address and cluster role of this node
* ``cluster.yaml`` - the state of the cluster, as seen by this dqlite node.
  It includes the same information as info.yaml, but for all cluster nodes.
* ``00000abcxx-00000abcxx``, ``open-abc`` - database segments
* ``cluster.crt``, ``cluster.key`` - node certificates
* ``snapshot-abc-abc-abc.meta``
* ``metadata{1,2}``
* ``*.sock`` - control unix sockets

> K8sd, which leverages the microcluster library, additionally contains:

Author: I'm dropping the Microcluster reference as per @louiseschmidtgen's suggestion.

* ``server.crt``, ``server.key`` certificates
* ``truststore`` folder, containing trusted certificates
* ``daemon.yaml`` - k8sd daemon configuration
* separate ``database`` folder

Dqlite cluster members have one of the following roles:

* 0 (voter) - Voting member of the database. A majority of voters is a quorum.
* 1 (stand-by) - Non-voting member of the database; can be promoted to voter.
* 2 (spare) - Not a member of the database.

> ## Stop Canonical K8s services on all nodes

Author: Just a minor nit: I saw that the other docs use ``{{product}}`` instead of ``{{ product }}``, so I did the same for consistency.

Before attempting to recover the cluster, all remaining Canonical K8s services
must be stopped. Use the following command on every node:

```
sudo snap stop k8s
```

## Reconfiguring the database

Out of the remaining cluster nodes that are still alive, we must pick the
one that has the most recent version of the Raft log.

> We need to update the ``cluster.yaml`` files, changing the role of the lost

Author: Should we use bold or italic? It improves readability, here's an example from the OpenStack docs: https://docs.openstack.org/nova/latest/configuration/index.html

nodes to "spare" (2). Additionally, double check the addresses and IDs specified
in ``cluster.yaml``, ``info.yaml`` and ``daemon.yaml``, especially if database
files were moved across nodes.

The following command can guide us through the process, prompting a text editor
with informative inline comments for each of the dqlite configuration file.

```
sudo /snap/k8s/current/bin/k8sd cluster-recover \
    --state-dir=/var/snap/k8s/common/var/lib/k8sd/state \
    --k8s-dqlite-state-dir=/var/snap/k8s/common/var/lib/k8s-dqlite \
    --log-level 0
```

Feel free to increase the log level for additional debug messages. Also note
that the command creates database backups before making any changes.

Once the Dqlite configuration is updated, the above command will reconfigure
the Raft members.

Finally, the command creates recovery tarballs that can be used to restore the
lost nodes.

Once the "cluster-recover" command completes, go ahead and restart the k8s
services on this node:

```
sudo snap start k8s
```

Make sure that the services started successfully by using ``sudo snap services k8s``.

> In a few moments, the ``k8s status`` command will report the cluster as ready.

Reviewer: nit: you can use `--wait-ready` to actually let the command wait until the node is ready.

## Recover the remaining nodes

The k8s-dqlite and k8sd recovery tarballs need to be copied over to all cluster
nodes.

For k8sd, all we have to do is copy ``recovery_db.tar.gz`` to
``/var/snap/k8s/common/var/lib/k8sd/state/recovery_db.tar.gz``. When the k8sd
service starts, it will load the archive and perform the necessary recovery steps.

The k8s-dqlite archive needs to be extracted manually. First, we'll create a backup
of the current k8s-dqlite state directory:

```
sudo mv /var/snap/k8s/common/var/lib/k8s-dqlite /var/snap/k8s/common/var/lib/k8s-dqlite.bkp
```

Next, we need to extract the backup archive:

```
sudo mkdir /var/snap/k8s/common/var/lib/k8s-dqlite
sudo tar xf  recovery-k8s-dqlite-$timestamp-post-recovery.tar.gz -C /var/snap/k8s/common/var/lib/k8s-dqlite
```

Node specific files need to be copied back to the k8s-dqlite state dir:

```
sudo cp /var/snap/k8s/common/var/lib/k8s-dqlite.bkp/cluster.crt /var/snap/k8s/common/var/lib/k8s-dqlite
sudo cp /var/snap/k8s/common/var/lib/k8s-dqlite.bkp/cluster.key /var/snap/k8s/common/var/lib/k8s-dqlite
sudo cp /var/snap/k8s/common/var/lib/k8s-dqlite.bkp/info.yaml /var/snap/k8s/common/var/lib/k8s-dqlite
```

> All set, we can now go ahead and restart the k8s services:

Reviewer 2: ```suggestion
Once these steps are completed, restart the k8s services:
```

```

> snap start k8s

Reviewer 2: ```suggestion
sudo snap start k8s
```

```

> Perform the same steps for the remaining nodes. Once a quorum is achieved, the cluster

Reviewer 2: ```suggestion
Repeat these steps for all remaining nodes. Once a quorum is achieved, the cluster
```
